### PR TITLE
Add llvm16 and clang16

### PR DESCRIPTION
**Pull request by dlorenc**
Fixes:

Related:

### Pre-review Checklist

cc @kaniini what kinda slotting do we need here?

### CODE REVIEW

**Comment by dlorenc** [comment]
image

**Comment by dlorenc** [comment]
It built!

File: llvm16.yaml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            name=${path##*/}
+            ln -s ../lib/llvm16/lib/$name "${{targets.destdir}}"/usr/lib/$name
+          done
+
+          mkdir -p "${{targets.destdir}}"/usr/lib/
+          for path in "${{targets.destdir}}"/usr/lib/llvm16/lib/*.so*; do
+            name=${path##*/}
+            ln -s ../lib/llvm16/lib/$name "${{targets.destdir}}"/usr/lib/$name
+          done
+      - runs: |
```
Comment (kaniini) [inline llvm16.yaml:108]:
```suggestion
      - working-directory: ${{targets.destdir}}/usr/lib/llvm16
        runs: |
```

**Review by imjasonh — DISMISSED** [review]

**Comment by dlorenc** [comment]
I've verified these packages can build Zig in #4182

File: clang-16.yaml [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+          mv "${{targets.destdir}}"/usr/share/scan-* "${{targets.subpkgdir}}"/usr/share/
+
+          mkdir -p "${{targets.subpkgdir}}"/usr/lib
+          mv "${{targets.destdir}}"/usr/lib/libear* "${{targets.subpkgdir}}"/usr/lib/
+    dependencies:
+      runtime:
+        - perl
+        - python3
+
+  - name: "clang-extras"
```
Comment (rawlingsj) [inline clang-16.yaml:159]: I think we also need `-16` suffix so we don't clash with https://github.com/wolfi-dev/os/blob/48479a25be862435f1420065cbdc677ae864272b/clang-15.yaml#L160

Comment (dlorenc) [reply clang-16.yaml:159]: Good catch. We should probably switch this to be slotted in the future.

File: clang-16.yaml [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      - runs: |
+          sonamefull="libclang-cpp.so.16"
+
+          mkdir -p "${{targets.subpkgdir}}"/usr/lib/
+          mv "${{targets.destdir}}"/usr/lib/$sonamefull "${{targets.subpkgdir}}"/usr/lib/
+    dependencies:
+      runtime:
+        - libLLVM-16
+
+  - name: "clang-analyzer"
```
Comment (rawlingsj) [inline clang-16.yaml:135]:
```suggestion
  - name: "clang-16-analyzer"
```

File: clang-16.yaml [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+          mkdir -p "${{targets.subpkgdir}}"/usr/share/clang
+          mv "${{targets.destdir}}"/usr/share/clang/clang-include-fixer* "${{targets.subpkgdir}}"/usr/share/clang/
+          mv "${{targets.destdir}}"/usr/share/clang/clang-format*.py "${{targets.subpkgdir}}"/usr/share/clang/
+          mv "${{targets.destdir}}"/usr/share/clang/*clang-tidy* "${{targets.subpkgdir}}"/usr/share/clang/
+          mv "${{targets.destdir}}"/usr/share/clang/run-find-all-symbols.py "${{targets.subpkgdir}}"/usr/share/clang/
+
+          mv "${{targets.destdir}}"/usr/share/emacs "${{targets.subpkgdir}}"/usr/share/
+
+  - name: "py3-clang"
```
Comment (rawlingsj) [inline clang-16.yaml:199]:
```suggestion
  - name: "py3-clang-16"
```

**Review by rawlingsj — APPROVED** [review]

**Pull request merged**